File: frontend/src/utils/mathText.js
/**
 * 数学文本分段与 LaTeX 化（供 MathText.vue 使用）。
 *
 * 设计目标：OCR/手输的自由文本里，把「分数 17/2、上标 x^2、\frac{}{}、\sqrt{}」
 * 这类数学片段识别出来交给 KaTeX 渲染，其余内容原样保留。
 * 原则：宁可漏渲染（退化为可读文本），不可误渲染（日期/比分/题号被拆成分数）。
 *
 * 三层防线（对应 OCR 输出风格不稳定的现实）：
 * 1. normalizeMathSource —— 剥掉 $…$、\( \) 这类定界符与纯排版命令，
 *    历史脏数据（如库里已存的 `\(|-\frac{5}{2}|\)`）也能就地洗干净；
 * 2. buildSegments —— 挑出可渲染的数学片段（含 ^{...} 带花括号上标）；
 * 3. texToReadable —— KaTeX 渲染失败时降级成「看得懂的文字」，
 *    绝不再把 \frac{1}{2} 这种源码直接摊给孩子看。
 */

const SIMPLE_EXPR_RE = '[A-Za-z0-9]+(?:[+\\-*/][A-Za-z0-9]+)*'
const PAREN_BASE_RE = `[（(](?:[^{}()（）\\r\\n]|\\{[^{}]*\\})+[）)]`

// 单个数学 token：
//  1) 上标：优先匹配带底数的幂（支持字母连写与中英文括号底数，含括号内含分数/负数如 (-\frac{3}{2})^2、(cd)^2027）
//  2) 已是 LaTeX 的命令：\frac{a}{b}、\sqrt{2}、\pi、\times …（命令后可跟一组花括号参数）
//  3) 斜杠分数：17/2（分子/分母各 1~3 位数字）
const TOKEN_RE = new RegExp(
  [
    `([0-9]+|[A-Za-z]+|${PAREN_BASE_RE})\\^\\s*(\\{[^{}]*\\}|\\([^)]*\\)|[-+]?[A-Za-z0-9]+)`, // 上标（优先匹配，避免被 \frac 拆碎）
    '\\\\[a-zA-Z]+\\s*(?:\\{[^{}]*\\})*', // LaTeX 命令（可带花括号参数）
    '([0-9]{1,3})/([0-9]{1,3})', // 斜杠分数（前面紧贴数字的情况在代码里排除，避免用 lookbehind 兼容旧 iOS Safari）
  ].join('|'),
  'g',
)

// 手机键盘、OCR 与聊天复制常把指数写为 Unicode 上标；先归一为 ^，再走同一渲染链路。
const UNICODE_SUPERSCRIPT_RE = new RegExp(`(${PAREN_BASE_RE}|[0-9A-Za-z]+)([⁰¹²³⁴⁵⁶⁷⁸⁹⁺⁻]+)`, 'g')
const UNICODE_SUPERSCRIPT_MAP = {
  '⁰': '0', '¹': '1', '²': '2', '³': '3', '⁴': '4',
  '⁵': '5', '⁶': '6', '⁷': '7', '⁸': '8', '⁹': '9',
  '⁺': '+', '⁻': '-',
}

// 分数黑名单：紧跟/前缀这些字样时视为日期、题号、比分等，不转分数
const FRAC_CTX_DENY = /[年月日时分秒页题号名楼第班次：%:]/i

// 尺寸/排版命令：只影响 KaTeX 排版，去掉不影响语义
const LAYOUT_CMDS = [
  'left', 'right',
  'bigl', 'bigr', 'Bigl', 'Bigr', 'bigg', 'Bigg', 'big', 'Big',
  'displaystyle', 'textstyle', 'limits', 'nolimits',
]

function isSafeFraction(num, den, raw, index, len) {
  if (Number(den) === 0) return false // 分母 0 必假
  // 前后各看 2 个字符：「第 23/24 题」「用时 9/12 分钟」这类语境不转
  const before = raw.slice(Math.max(0, index - 2), index)
  const after = raw.slice(index + len, index + len + 2)
  if (/[0-9]/.test(before.slice(-1))) return false // 1234/5 只会匹配到 234/5 残串，直接放弃
  if (FRAC_CTX_DENY.test(before) || FRAC_CTX_DENY.test(after)) return false
  // 连续多段斜杠（2024/9/12 这类日期串）不转
  if (before.endsWith('/') || after.startsWith('/')) return false
  return true
}

function toSuperscriptTex(base, expRaw) {
  // x^2 → x^{2}；（cd）^2027 → (cd)^{2027}；x^(2n-1) → x^{2n-1}
  let exp = expRaw.trim()
  if (exp.startsWith('{') && exp.endsWith('}')) exp = exp.slice(1, -1)
  if (exp.startsWith('(') && exp.endsWith(')')) exp = exp.slice(1, -1)
  if (!new RegExp(`^${SIMPLE_EXPR_RE}$`).test(exp)) return null // 含其它符号则不转
  return `${base.replace(/[（]/g, '(').replace(/[）]/g, ')')}^{${exp}}`
}

function normalizeUnicodeSuperscripts(s) {
  return s.replace(UNICODE_SUPERSCRIPT_RE, (_, base, superscript) => {
    const exponent = [...superscript].map((ch) => UNICODE_SUPERSCRIPT_MAP[ch] || ch).join('')
    return `${base.replace(/[（]/g, '(').replace(/[）]/g, ')')}^${exponent}`
  })
}

/**
 * 清洗 OCR/手输文本里的 LaTeX 定界符与纯排版命令。
 *
 * 云端视觉模型会自由选择输出风格：纯文本 / 裸 LaTeX / 带定界符的 LaTeX。
 * 定界符（$、$$、\(、\)、\[、\]）不是本渲染器的数学边界，不剥掉就会作为
 * 可见字符留在题干里（孩子会看到「$」「\」）。后端入库前已清洗一遍，
 * 这里再兜一次，保证历史记录与手工粘贴的内容同样干净。
 */
export function normalizeMathSource(raw) {
  if (!raw) return ''
  let s = String(raw)

  // 换行命令 \\ 还原成真实换行（必须在清理反斜杠之前）
  s = s.replace(/\\\\/g, '\n')
  // 转义字面符号
  s = s.replace(/\\\{/g, '{').replace(/\\\}/g, '}')
    .replace(/\\_/g, '_').replace(/\\%/g, '%')
    .replace(/\\&/g, '&').replace(/\\#/g, '#')
    .replace(/\\\$/g, '')
    .replace(/\\(?:,|;|:|!| )/g, ' ')
  // 定界符：只删符号本身，内容原样保留
  s = s.replace(/\$\$/g, '').replace(/\$/g, '')
  s = s.replace(/\\\[/g, '').replace(/\\\]/g, '')
  s = s.replace(/\\\(/g, '').replace(/\\\)/g, '')
  // \text{...} / \mathrm{...} 这类「内容即正文」的包装只留内容
  s = s.replace(/\\(?:text|mathrm|mathbf|mathit|mathbb|mathcal|operatorname)\s*\{([^{}]*)\}/g, '$1')
  // 尺寸/排版命令与空白命令
  LAYOUT_CMDS.forEach((cmd) => { s = s.replace(new RegExp('\\\\' + cmd + '(?![a-zA-Z])', 'g'), '') })
  s = s.replace(/\\(?:quad|qquad|enspace|thinspace|medspace|thickspace)(?![a-zA-Z])/g, ' ')
  s = normalizeUnicodeSuperscripts(s)

  return s
}

/**
 * 把自由文本切成 [{type:'text'|'math', value}] 段。
 * math 段的 value 是可直接喂给 KaTeX 的 LaTeX 源串。
 */
export function buildSegments(raw) {
  const out = []
  const source = normalizeMathSource(raw)
  if (!source) return out
  let last = 0
  TOKEN_RE.lastIndex = 0
  let m
  while ((m = TOKEN_RE.exec(source)) !== null) {
    let tex = null
    if (m[1] !== undefined) {
      tex = toSuperscriptTex(m[1], m[2])
      if (!tex) continue
    } else if (m[0].startsWith('\\')) {
      tex = m[0].replace(/\s+/, '') // \frac {17}{2} → \frac{17}{2}
    } else if (m[3] !== undefined) {
      // 斜杠分数：过黑名单才转
      if (!isSafeFraction(m[3], m[4], source, m.index, m[0].length)) continue
      tex = `\\frac{${m[3]}}{${m[4]}}`
    }
    if (!tex) continue
    if (m.index > last) out.push({ type: 'text', value: source.slice(last, m.index) })
    out.push({ type: 'math', value: tex })
    last = m.index + m[0].length
  }
  if (last < source.length) out.push({ type: 'text', value: source.slice(last) })
  return out
}

export function escapeHtml(s) {
  return s.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;')
}

function replaceCmd(s, name, ch) {
  return s.replace(new RegExp('\\\\' + name + '(?![a-zA-Z])', 'g'), ch)
}

/**
 * KaTeX 渲染失败时的降级：把 LaTeX 转成「看得懂的中文数学写法」。
 *
 * 为什么不直接显示源码：`\begin{cases}` / `\left` 缺配对等情况会让 KaTeX 抛错，
 * 此时把 `\frac{1}{2}` 摊给孩子看毫无意义，不如退成 `(1)/(2)`。
 */
export function texToReadable(tex) {
  let s = String(tex || '')
  s = s.replace(/\\begin\{[^}]*\}/g, '').replace(/\\end\{[^}]*\}/g, '')
  s = s.replace(/\\(?:d|t|c)?frac\s*\{([^{}]*)\}\s*\{([^{}]*)\}/g, '($1)/($2)')
  s = s.replace(/\\sqrt\s*\{([^{}]*)\}/g, '√($1)')
  s = s.replace(/\^\{([^{}]*)\}/g, '^$1')
  const sym = {
    times: '×', div: '÷', cdot: '·', pm: '±', mp: '∓',
    leq: '≤', le: '≤', geq: '≥', ge: '≥', neq: '≠', ne: '≠',
    angle: '∠', triangle: '△', parallel: '∥', perp: '⊥',
    pi: 'π', alpha: 'α', beta: 'β', gamma: 'γ', theta: 'θ',
    lambda: 'λ', mu: 'μ', sigma: 'σ', omega: 'ω', Delta: 'Δ',
    infty: '∞', degree: '°',
  }
  Object.keys(sym).forEach((k) => { s = replaceCmd(s, k, sym[k]) })
  s = s.replace(/[{}]/g, '')
  s = s.replace(/\\/g, '')
  return s.trim()
}

// KaTeX 惰性加载：主包不膨胀，首次遇到数学片段才拉取独立 chunk
let katexPromise = null
function loadKatex() {
  if (!katexPromise) katexPromise = import('katex').then((m) => m.default || m)
  return katexPromise
}

/**
 * 渲染为 HTML 字符串（同步返回兜底版；KaTeX 就绪后通过 upgrade 回调升级）。
 * 返回 { html, hasMath, upgrade } —— upgrade() 为 null 表示无需升级。
 */
export function renderMathHtml(raw) {
  const segs = buildSegments(raw || '')
  if (segs.length === 0) return { html: '', hasMath: false, upgrade: null }
  const hasMath = segs.some((s) => s.type === 'math')
  // KaTeX 就绪前的兜底版同样不暴露 LaTeX 源码，避免首帧闪出「\frac」
  const plain = segs
    .map((s) => escapeHtml(s.type === 'math' ? texToReadable(s.value) : s.value))
    .join('')
  if (!hasMath) return { html: plain, hasMath: false, upgrade: null }

  const upgrade = async () => {
    const katex = await loadKatex()
    return segs
      .map((s) => {
        if (s.type !== 'math') return escapeHtml(s.value)
        try {
          return katex.renderToString(s.value, { throwOnError: true, strict: false })
        } catch {
          // 渲染失败：退成可读数学写法，不留 LaTeX 源码
          return `<span class="math-fallback">${escapeHtml(texToReadable(s.value))}</span>`
        }
      })
      .join('')
  }
  return { html: plain, hasMath: true, upgrade }
}

File: backend/app/utils/math_text.py
"""数学文本规范化：把 OCR 产出的 LaTeX 变回「前端渲染器吃得下」的干净文本。

为什么需要它
------------
云端视觉模型（glm-4v-flash 等）输出风格不稳定，同一张数学题照片可能给出三种形态：

1. 纯文本          ``x = (-b + sqrt(b^2-4ac)) / 2a``
2. 裸 LaTeX 片段   ``\\sqrt{18} - 3\\sqrt{2} + \\frac{1}{2}``
3. 带定界符的 LaTeX ``\\(|-\\frac{5}{2}|\\)``、``$\\frac{2}{x}=1$``

前端 ``MathText`` 只认「裸 LaTeX 片段」，定界符（``$``、``$$``、``\\(``、``\\)``、``\\[``、``\\]``）
不会被识别成数学边界，于是作为可见字符残留在题干里 —— 孩子会看到「$」「\\」。
这里在**入库之前**统一剥掉定界符、清理纯排版命令、并让真正无法渲染的命令
至少不裸露反斜杠，保证错题卡片、编辑框、复习卷打印三处都是干净的。
"""

from __future__ import annotations

import re

# 尺寸/排版类命令：只是给 KaTeX 看的，去掉不影响语义（\left| → |、\displaystyle → 空）
_LAYOUT_CMDS = (
    "left",
    "right",
    "big",
    "Big",
    "bigg",
    "Bigg",
    "bigl",
    "bigr",
    "Bigl",
    "Bigr",
    "displaystyle",
    "textstyle",
    "scriptstyle",
    "scriptscriptstyle",
    "limits",
    "nolimits",
)

# 带花括号参数、但内容本身就是正文的命令：\text{厘米} → 厘米
_TEXT_WRAPPER_CMDS = ("text", "mathrm", "mathbf", "mathit", "mathbb", "mathcal", "operatorname")

# 空白类命令：替换成普通空格
_SPACE_CMDS = (
    "quad",
    "qquad",
    "hspace",
    "enspace",
    "thinspace",
    "medspace",
    "thickspace",
    "negthinspace",
)

# KaTeX 能渲染、且初中数学题干里会出现的命令（保留，绝不剥掉反斜杠）
_KATEX_KEEP = {
    # 分式与根式
    "frac", "dfrac", "tfrac", "cfrac", "sqrt", "binom", "choose",
    # 运算与关系
    "times", "div", "cdot", "pm", "mp", "ast", "star", "circ", "bullet",
    "le", "leq", "ge", "geq", "ne", "neq", "lt", "gt", "equiv", "approx", "sim", "simeq", "cong",
    "propto", "because", "therefore", "implies", "iff", "land", "lor", "neg", "forall", "exists",
    # 几何与集合
    "angle", "measuredangle", "degree", "triangle", "square", "parallel", "nparallel", "perp",
    "in", "notin", "subset", "subseteq", "supset", "cup", "cap", "emptyset", "varnothing", "mid",
    "lfloor", "rfloor", "lceil", "rceil", "langle", "rangle", "vert", "Vert",
    # 微积分/数列（初中偶见）
    "sum", "prod", "int", "infty", "lim", "partial", "nabla",
    # 函数与对数
    "log", "ln", "lg", "sin", "cos", "tan", "cot", "sec", "csc", "arcsin", "arccos", "arctan",
    # 希腊字母
    "alpha", "beta", "gamma", "delta", "epsilon", "varepsilon", "zeta", "eta", "theta", "vartheta",
    "iota", "kappa", "lambda", "mu", "nu", "xi", "rho", "sigma", "tau", "upsilon", "phi", "varphi",
    "chi", "psi", "omega", "Gamma", "Delta", "Theta", "Lambda", "Xi", "Pi", "Sigma", "Phi", "Psi", "Omega",
    # 箭头
    "to", "mapsto", "leftarrow", "rightarrow", "leftrightarrow", "uparrow", "downarrow", "updownarrow",
    "Leftarrow", "Rightarrow", "Leftrightarrow", "longleftarrow", "longrightarrow", "longleftrightarrow",
    # 修饰
    "overline", "underline", "vec", "bar", "hat", "tilde", "widehat", "widetilde", "dot", "ddot",
    "overset", "underset", "stackrel", "boxed", "prime", "backprime", "pmod", "bmod", "mod",
    # 省略号
    "dots", "ldots", "cdots", "vdots", "ddots", "dotsb", "dotsc",
    # 环境（\begin{cases} 等靠前端 KaTeX 兜底）
    "begin", "end",
}

# 转义的字面符号：剥掉反斜杠、留下符号本身
_ESCAPED_LITERALS = {
    "\\{": "{",
    "\\}": "}",
    "\\_": "_",
    "\\%": "%",
    "\\&": "&",
    "\\#": "#",
    "\\$": "",
    "\\ ": " ",
    "\\,": " ",
    "\\;": " ",
    "\\:": " ",
    "\\!": "",
}

# 剩余的反斜杠命令：保留白名单，其余只留命令名（不让「\」露给孩子看）
_CMD_RE = re.compile(r"\\([a-zA-Z]+)")

# 手机键盘、OCR 与聊天复制常使用 Unicode 上标（m²、x⁻²）；统一成前端
# MathText 与 KaTeX 都能稳定处理的 ^ 写法。只转换 ASCII 数学底数，避免误伤中文正文。
_UNICODE_SUPERSCRIPT_RE = re.compile(
    r"([A-Za-z0-9]+|[（(][A-Za-z0-9]+(?:[+\-*/][A-Za-z0-9]+)*[）)])([⁰¹²³⁴⁵⁶⁷⁸⁹⁺⁻]+)"
)
_UNICODE_SUPERSCRIPT_MAP = str.maketrans({
    "⁰": "0", "¹": "1", "²": "2", "³": "3", "⁴": "4",
    "⁵": "5", "⁶": "6", "⁷": "7", "⁸": "8", "⁹": "9",
    "⁺": "+", "⁻": "-",
})


def _normalize_unicode_superscripts(text: str) -> str:
    """将 OCR 常见的 Unicode 上标统一为 ^，供前端 KaTeX 稳定渲染。"""
    def _replace(match: re.Match) -> str:
        base = match.group(1).replace("（", "(").replace("）", ")")
        return f"{base}^{match.group(2).translate(_UNICODE_SUPERSCRIPT_MAP)}"

    return _UNICODE_SUPERSCRIPT_RE.sub(_replace, text)


def _strip_wrappers(text: str) -> str:
    """把 \\text{...} 这类「内容即正文」的包装剥掉，保留花括号内的内容。"""
    for cmd in _TEXT_WRAPPER_CMDS:
        pattern = re.compile(r"\\" + cmd + r"\s*\{([^{}]*)\}")
        prev = None
        while prev != text:  # 允许嵌套，反复收敛
            prev = text
            text = pattern.sub(r"\1", text)
    return text


def normalize_math_text(text: str | None) -> str:
    """清洗 OCR 输出的数学文本，返回可直接入库/渲染的纯文本。

    幂等：重复调用结果不变。
    """
    if not text:
        return ""

    out = text

    # 1) 换行命令 \\ 还原成真实换行（必须在清理反斜杠之前做）
    out = out.replace("\\\\", "\n")

    # 2) 转义字面符号（\$ 直接丢弃，题干里没有合法的美元符号语义）
    for src, dst in _ESCAPED_LITERALS.items():
        out = out.replace(src, dst)

    # 3) 定界符：$$…$$ / $…$ / \( \) / \[ \] 全部剥掉，只留内部内容
    out = out.replace("$$", "")
    out = out.replace("\\[", "").replace("\\]", "")
    out = out.replace("\\(", "").replace("\\)", "")
    out = out.replace("$", "")

    # 4) 正文包装命令
    out = _strip_wrappers(out)

    # 5) 尺寸/排版命令与空白命令
    for cmd in _LAYOUT_CMDS:
        out = re.sub(r"\\" + cmd + r"(?![a-zA-Z])", "", out)
    for cmd in _SPACE_CMDS:
        out = re.sub(r"\\" + cmd + r"(?![a-zA-Z])", " ", out)

    # 5.1) OCR / 手机输入的 Unicode 上标统一成 ^ 形式
    out = _normalize_unicode_superscripts(out)

    # 6) 白名单之外的反斜杠命令：只丢反斜杠，命令名留作普通文字，避免裸「\」进入题干
    def _keep_or_strip(m: re.Match) -> str:
        name = m.group(1)
        return m.group(0) if name in _KATEX_KEEP else name

    out = _CMD_RE.sub(_keep_or_strip, out)

    # 7) 收敛空白：行内多余空格压成一个，行首行尾去空；保留换行结构
    lines = [re.sub(r"[ \t\u3000]{2,}", " ", ln).strip() for ln in out.split("\n")]
    return "\n".join(lines).strip()
